### PR TITLE
Add type hints for tf models final batch

### PR DESCRIPTION
**Pull request by nablabits**
# What does this PR do?

Addresses, hopefully :crossed_fingers: , the last the type hints for tf models in https://github.com/huggingface/transformers/issues/16059:

1. `RegNet`
	1. `TFRegNetForImageClassification`
	2. `TFRegNetModel`
2. `TFSamModel`
3. `TFSegformerDecodeHead`
4. `TransfoXL`
	1. `TFTransfoXLLMHeadModel`
	2. `TFTransfoXLModel`
5. `TFWav2Vec2ForSequenceClassification`
6. `TFXLMModel`

## Who can review?
@Rocketknight1 please

### CODE REVIEW

**Comment by nablabits** [comment]
There are some tests failing likely because they are running on python 3.8.12 and the `|` notation for type hints was introduced in python 3.10:

image

It seems that it's only the [RegNet model](https://github.com/huggingface/transformers/blob/f353448a4dd4fb412be789a863ff6cabb3f96c70/src/transformers/models/regnet/modeling_tf_regnet.py#L380), so I will replace it.

It's not clear to me, though, why the others are not complaining :thinking:  
Anyways, I assume that eventually the circle ci version will be bumped and that will make them compliant in terms of type hints.

File: src/transformers/models/xlm/modeling_tf_xlm.py [hunk 1]
```diff
@@ -700,20 +699,20 @@ def __init__(self, config, *inputs, **kwargs):
...
-        output_hidden_states=None,
-        return_dict=None,
-        training=False,
-    ) -> Union[TFBaseModelOutput, Tuple[tf.Tensor]]:
+        input_ids: TFModelInputType | None = None,
+        attention_mask: np.ndarray | tf.Tensor | None = None,
+        langs: np.ndarray | tf.Tensor | None = None,
+        token_type_ids: np.ndarray | tf.Tensor | None = None,
+        position_ids: np.ndarray | tf.Tensor | None = None,
+        lengths: np.ndarray | tf.Tensor | None = None,
```
Comment (Rocketknight1) [inline src/transformers/models/xlm/modeling_tf_xlm.py:707]: I think it's okay to just say `tf.Tensor` here, even though `np.ndarray` inputs are accepted and implicitly converted too.

File: src/transformers/models/regnet/modeling_tf_regnet.py [hunk 2]
```diff
@@ -447,11 +447,11 @@ def __init__(self, config: RegNetConfig, *inputs, **kwargs):
...
         self,
-        pixel_values: tf.Tensor = None,
-        labels: tf.Tensor = None,
-        output_hidden_states: bool = None,
-        return_dict: bool = None,
-        training=False,
+        pixel_values: Optional[tf.Tensor] = None,
+        labels: Optional[tf.Tensor] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/regnet/modeling_tf_regnet.py:453]: Maybe do `tf.Tensor | None` here to match the rest?

**Review by Rocketknight1 — APPROVED** [review]
Woah, I never thought we'd actually be able to finish the type hints list. This is amazing! I added a couple of optional comments, but other than that this is good to merge. Let me know whenever you're ready!

File: src/transformers/models/regnet/modeling_tf_regnet.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/models/regnet/modeling_tf_regnet.py:453]: Ah, I just saw your comment above! Our CI runner should be able to handle `| None` - I wonder what went wrong?

File: src/transformers/models/xlm/modeling_tf_xlm.py (hunk 1 above)
Comment (nablabits) [reply src/transformers/models/xlm/modeling_tf_xlm.py:707]: Yeah, I don't have a strong opinion here as, to the top off my head, there's a nice mix of these through the TF models, some just show `tf.Tensor` and some others show the `ndarray` bit. 
Specifically, in this model, all the annotated `call`s are with the latter so for consistency we should leave it but then we will be falling short in the overall consistency that, as you say, `np.ndarray` are implicitly converted :thinking:

File: src/transformers/models/regnet/modeling_tf_regnet.py (hunk 2 above)
Comment (nablabits) [reply src/transformers/models/regnet/modeling_tf_regnet.py:453]: Yeah, I find it quite uncanny, as the other annotations are not complaining :thinking: 

I'm happy to further investigate this, are you happy for me to open a follow up issue, so we can close this dreary type hints issue, or do you think that there are more pressing stuff to be solved before (the list of issues is already huge)?

**Comment by nablabits** [comment]
> Woah, I never thought we'd actually be able to finish the type hints list. This is amazing! I added a couple of optional comments, but other than that this is good to merge. Let me know whenever you're ready!

Yeah, it's the magic of perseverance :sparkles: 

I'm happy for you to merge, let me know if you want me to open a follow up issue for the [circle ci thingy](https://github.com/huggingface/transformers/pull/25883#discussion_r1312612246)

File: src/transformers/models/regnet/modeling_tf_regnet.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/models/regnet/modeling_tf_regnet.py:453]: I don't think there's any real need here, since both kinds of type hints are fine!

**Comment by Rocketknight1** [comment]
Type hinting project is officially complete as of this PR!

**Pull request merged**